mapper/collectd_mapper/src/batcher.rs

```diff
@@ -1,169 +1,171 @@
+        // which might in the worst case end up in a busy loop.
+        let delta = next_notification_at - self.clock.now();
+        let t20ms = std::time::Duration::from_millis(20);
+        let next_notification_in = std::cmp::max(t20ms, delta.to_std().unwrap_or(t20ms));
 
-        // Creates a sleep timer future handler and does not await here
-        // for sleep to finish, but inside the select loop
-        let sleep = time::sleep(self.batching_window);
-        tokio::pin!(sleep);
+        next_notification_in
+    }
 
-        loop {
-            select! {
-                _ = &mut sleep => {
-                        break;
-                }
-                maybe_message = self.receive_message(messages) => {
-                    match maybe_message {
-                        Some((message, _timestamp)) => {
-                            let collectd_message = match CollectdMessage::parse_from(&message) {
-                                Ok(message) => message,
-                                Err(err) => {
-                                    error!("Error parsing collectd message: {}", err);
-                                    continue;   // Even if one message is faulty, we skip that one and keep building the batch
-                                },
-                            };
-                            message_batch.add_to_batch(collectd_message)?;
-                        }
-                        None => break
+    async fn process_io(
+        &mut self,
+        messages: &mut dyn MqttMessageStream,
+        next_notify_in: std::time::Duration,
+        outputs: &mut Vec<message_batcher::Output<OwnedCollectdMessage>>,
+    ) {
+        match tokio::time::timeout_at(
+            tokio::time::Instant::now() + next_notify_in,
+            messages.next(),
+        )
+        .await
+        {
+            Err(_) => {
+                // Timeout fired. Inform functional core
+                let now = self.clock.now();
+                self.batcher
+                    .handle(message_batcher::Input::Notify { now }, outputs);
+            }
+            Ok(Some(mqtt_message)) => {
+                // got a message
+                let received_at = self.clock.now();
+                match CollectdMessage::parse_from(&mqtt_message) {
+                    Ok(collectd_message) => self.batcher.handle(
+                        message_batcher::Input::Message {
+                            received_at,
+                            message: collectd_message.into(),
+                        },
+                        outputs,
+                    ),
+                    Err(err) => {
+                        error!("Error parsing collectd message: {}", err);
                     }
                 }
-
+            }
+            Ok(None) => {
+                // XXX: Not sure if this works!
+                error!("MQTT connection closed. Retrying...");
```

Comment:
Indeed. One receives `Ok(None)` when the client itself closes the connection. When the connection is lost, the mqtt crate tries to re-establish the connection and this channel is not closed. I would simply exit here.